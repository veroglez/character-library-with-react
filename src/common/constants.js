export default {
  DOMAIN: 'https://rickandmortyapi.com/api/',
  DOMAIN_CHARACTER: 'https://rickandmortyapi.com/api/character',
};
