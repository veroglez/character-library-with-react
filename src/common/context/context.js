import { node } from 'prop-types';
import React, { createContext, Component } from 'react';
import { C, L10N } from '..';

const Context = createContext();
const { Provider, Consumer: ConsumerData } = Context;

class ProviderData extends Component {
  static propTypes = {
    children: node,
  };

  static defaultProps = {
    children: undefined,
  };

  state = {
    next: '',
    characters: [],
    l10n: L10N,
  }

  componentDidMount() {
    const service = 'character';
    this.onFetch(service);
  }

  onFetch = async (service, next) => {
    const { characters } = this.state;
    const response = await fetch(next || `${C.DOMAIN}${service}`);
    const character = await response.json();
    this.setState({ characters: next ? [...characters, ...character.results] : character.results, next: character.info.next });
  }

  onData = object => this.setState(object);

  render() {
    const {
      onData, onFetch, props: { children },
    } = this;

    return (
      <Provider value={{ ...this.state, onData, onFetch }}>
        { children }
      </Provider>
    );
  }
}

export { ConsumerData, ProviderData };
