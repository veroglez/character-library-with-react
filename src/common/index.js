import C from './constants';
import L10N from './l10n';

export {
  C,
  L10N,
};
