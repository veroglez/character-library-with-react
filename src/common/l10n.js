export default {
  CHARACTERS: 'Characters',
  GENDER: 'Gender',
  ORIGIN: 'Origin',
  SEARCH: 'Search...',
  STATUS: 'Status',
};
