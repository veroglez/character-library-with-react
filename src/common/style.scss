$color: (
  dark: #222222,
  light: #e3e3e3,
  rnm-yellow: #fde145,
  rnm-green: #30c52e,
  rnm-orange: #fe7250,
);

$breakpoint: (
  xs: 320px,
  small: 600px,
  medium: 840px,
  large: 1024px,
);

@mixin mobileLarge { // > 321
  @media (min-width: #{map-get($breakpoint, xs)+1}) { @content; }
}
@mixin tablet { // > 601
  @media (min-width: #{map-get($breakpoint, small)+1}) { @content; }
}
@mixin tabletLarge {
  @media (min-width: #{map-get($breakpoint, medium)}) { @content; }
}
@mixin desktop {
  @media (min-width: #{map-get($breakpoint, large)}) { @content; }
}