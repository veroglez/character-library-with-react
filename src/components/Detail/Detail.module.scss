@import "../../common/style.scss";

.container {
  margin: 0px 20px 40px;
  position: relative;
  
  @include tablet {
    display: flex;
    margin: 0px 60px 40px;
  }

  .add {
    background-color: map-get($color, light);
    border-radius: 20px;
    padding: 8px; 
    margin-bottom: 10px;
    margin-top: 20px;
    cursor: pointer;
    text-align: center;
  }

  .bold {
    font-weight: 600;
  }
  
  .caption {
    padding: 6px 12px;
    margin-bottom: 20px;
    margin-top: 20px;

    @include tablet {
      margin-top: 0px;
    }
  }

  .image {
    width: 100%;
    height: 246px;
    margin-right: 40px;

    @include tablet {
      width: 308px;
    }
  }

  .list {
    margin-bottom: 12px;
  }

  .text {
    display: inline-block;
    margin-right: 8px;
  }

  .favourites {
    position: absolute;
    top: 10px;
    left: 10px;
  }
}