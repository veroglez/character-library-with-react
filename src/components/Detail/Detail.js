import { func, string, shape } from 'prop-types';
import React, { PureComponent } from 'react';
import { ConsumerData } from '../../common/context/context';
import scss from './Detail.module.scss';
import { Image, Text } from '..';
import { C } from '../../common';
import onColor from '../../common/modules/onColor';
import assets from '../../assets';


class Detail extends PureComponent {
  static propTypes = {
    characterId: string.isRequired,
    l10n: shape({}).isRequired,
    onTitle: func,
  };

  static defaultProps = {
    onTitle: undefined,
  };

  state = {
    image: '',
    name: undefined,
    gender: undefined,
    species: undefined,
    status: undefined,
    origin: undefined,
  }

  async componentDidMount() {
    const { characterId, onTitle } = this.props;
    const response = await fetch(`${C.DOMAIN_CHARACTER}/${characterId}`);
    const {
      id, image, name, gender, species, status, origin: { name: origin },
    } = await response.json();
    this.setState({
      id, image, name, gender, species, status, origin,
    });
    onTitle(name);
  }

  add = (id, favourites, onData) => {
    let data = [...new Set([...favourites.data, id])];
    onData({ favourites: { data } });
    data = JSON.stringify({ data });
    localStorage.setItem('rnm:favourites', data);
  }

  render() {
    const {
      add,
      state: {
        id, image, name, gender, species, status, origin,
      },
      props: { l10n },
    } = this;
    const info = { gender, status, origin };

    return (
      <ConsumerData>
        {({ favourites, onData }) => (
          <div className={scss.container}>
            {favourites.data.includes(id) && <Image icon image={assets.heart} styles={scss.favourites} />}
            <Image image={image} styles={scss.image} />
            <div className={scss.info}>
              <Text caption color={onColor(species)} styles={scss.caption}>{species}</Text>
              <Text headline>{name}</Text>
              <ul>
                {Object.keys(info).map((key, index) => (
                  <li key={index.toString()} className={scss.list}>
                    <Text styles={`${scss.text} ${scss.bold}`}>{`${l10n[key.toUpperCase()]}:`}</Text>
                    <Text styles={scss.text}>{info[key]}</Text>
                  </li>
                ))}
              </ul>
              <div role="presentation" onClick={() => add(id, favourites, onData)}>
                <Text styles={scss.add}>{l10n.ADD}</Text>
              </div>
            </div>
          </div>
        )}
      </ConsumerData>
    );
  }
}

export default Detail;
