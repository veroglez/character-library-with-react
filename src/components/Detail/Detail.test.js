import React from 'react';
import renderer from 'react-test-renderer';
import Detail from './Detail';
import { L10N } from '../../common';
import { ProviderData } from '../../common/context/context';

// const fetch = jest.fn(() => Promise.resolve());

const CHARACTER_ID = '1';
const FUNC = () => {};

const HOC = inherit => (
  <ProviderData favourites={[1, 2]} onData={() => {}}>
    <Detail {...inherit} />
  </ProviderData>
);


describe('<Detail>', () => {
  it('renders when {characterId} and {l10n}', async () => {
    const tree = renderer.create(<HOC characterId={CHARACTER_ID} l10n={L10N} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders when {onTitle}', async () => {
    const tree = renderer.create(<HOC characterId={CHARACTER_ID} l10n={L10N} onTitle={FUNC} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
