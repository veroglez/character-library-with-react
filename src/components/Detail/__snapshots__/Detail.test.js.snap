// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Detail> renders when {characterId} and {l10n} 1`] = `
<div
  className="container"
>
  <div
    className="image image"
    style={
      Object {
        "backgroundImage": "url()",
      }
    }
  />
  <div
    className="info"
  >
    <p
      className="text caption caption orange"
    />
    <p
      className="text headline"
    />
    <ul>
      <li
        className="list"
      >
        <p
          className="text body text bold"
        >
          Gender:
        </p>
        <p
          className="text body text"
        />
      </li>
      <li
        className="list"
      >
        <p
          className="text body text bold"
        >
          Status:
        </p>
        <p
          className="text body text"
        />
      </li>
      <li
        className="list"
      >
        <p
          className="text body text bold"
        >
          Origin:
        </p>
        <p
          className="text body text"
        />
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Detail> renders when {onTitle} 1`] = `
<div
  className="container"
>
  <div
    className="image image"
    style={
      Object {
        "backgroundImage": "url()",
      }
    }
  />
  <div
    className="info"
  >
    <p
      className="text caption caption orange"
    />
    <p
      className="text headline"
    />
    <ul>
      <li
        className="list"
      >
        <p
          className="text body text bold"
        >
          Gender:
        </p>
        <p
          className="text body text"
        />
      </li>
      <li
        className="list"
      >
        <p
          className="text body text bold"
        >
          Status:
        </p>
        <p
          className="text body text"
        />
      </li>
      <li
        className="list"
      >
        <p
          className="text body text bold"
        >
          Origin:
        </p>
        <p
          className="text body text"
        />
      </li>
    </ul>
  </div>
</div>
`;
